### ISSUE TITLE
AVT 2 - Modal React Component - keyboard navigation improvements

### ISSUE THREAD

**Issue by snidersd** [carbon-design-system/carbon#2979]
## Environment
macOS Mojave version 10.14.2
Chrome version 72.0.3626.96 (Official Build) (64-bit)
Carbon v10

## Detailed Description
After the modal opens the default focus is not on the primary button.
A mouse click outside the modal closes the modal.

## Steps to reproduce the issue
1. Click the button to open the modal. Focus is on the close button "X".
2. Click outside the modal and the modal closes.

## Expected Result
1. Initial focus should be on the primary button.
2. Focus should remain in the modal until it is closed.

See [ARIA Authoring Practices Guidelines Modal Design pattern](https://www.w3.org/TR/wai-aria-practices-1.1/examples/dialog-modal/dialog.html).

**Comment by emyarod** [comment]
> After the modal opens the default focus is not on the primary button

as a workaround this is configurable with the `selectorPrimaryFocus` prop, but I can modify the default behavior

> Focus should remain in the modal until it is closed

for this point, I just want to double check with design about this UX change first

**Comment by dakahn** [comment]
@snidersd looking [the authoring practices for Modal](https://www.w3.org/TR/wai-aria-practices-1.1/#dialog_modal) they suggest initial focus be placed on the _first focusable element_  unless you have a good reason not to. 🤔. Further can't there be issues with NVDA when the default first focusable element comes _after_ the dialogue's main content?

[reference here](https://developer.paciellogroup.com/blog/2018/06/the-current-state-of-modal-dialog-accessibility/) at section "Overly verbose NVDA announcements".

**Comment by dakahn** [comment]
Also @emyarod I assigned this to you. Figured that's what you meant by "...I can modify the default behavior". If not, let me know and this can be delegated. 👍🏽

**Comment by emyarod** [comment]
> 1. Initial focus should be on the primary button.
> 2. Focus should remain in the modal until it is closed.

the first point is only a docs change if anything, since the component already support changing the default primary focus element. I'm leaning towards this being a non issue

the second point is not actionable until we have confirmation from design

**Comment by snidersd** [comment]
@dakahn This issue should remain open.

**Comment by snidersd** [comment]
@dakahn This issue should remain open.

**Comment by emyarod** [comment]
> Initial focus should be on the primary button.

so it seems this point is a non issue according to the [modal dialog guidelines](https://www.w3.org/TR/wai-aria-practices-1.1/#dialog_modal), and we added a [prop for modifying the first focused element](https://github.com/carbon-design-system/carbon/commit/3bc0c379636e271a2db3cdfe76269e92e46f5361) a year prior to this ticket

> Focus should remain in the modal until it is closed.

to clarify, is this about focus trap behavior (which is already implemented) or is it explicitly about the default behavior of closing the modal upon clicking outside the dialog box?

**Comment by emyarod** [comment]
cc @snidersd @dakahn

**Comment by snidersd** [comment]
@emyarod @dakahn  The [WCAG 2.1 example](https://www.w3.org/WAI/GL/wiki/Using_ARIA_role%3Ddialog_to_implement_a_modal_dialog_box)and the [ARIA example](https://www.w3.org/TR/wai-aria-practices-1.1/examples/dialog-modal/dialog.html) do not allow the user to close the modal by clicking outside the modal, but examples of a [date picker dialog](https://www.w3.org/TR/wai-aria-practices-1.1/examples/dialog-modal/datepicker-dialog.html) allows the modal to close when there is a mouse click outside the modal. IMO the handling of a mouse click to close the modal is based on the purpose of the modal, but going by the WCAG and ARIA examples, the default should be to keep the focus inside the modal and allow the content author to change the interaction if needed.

**Comment by emyarod** [comment]
just want to confirm with @carbon-design-system/design that this is a UX change we want to make (not closing modals by default when clicking outside of the modal). currently we close the modal by default but only in `ComposedModal`s I believe. our default `Modal` component does not have this behavior

**Comment by aagonzales** [comment]
How a modal is closed depends on the modal type. We have it documented on the modal usage page. https://www.carbondesignsystem.com/components/modal/usage#passive-modal

Passive modals are the only modal type that should have the function to close by clicking outside the of modal window.

**Comment by emyarod** [comment]
thanks for the clarification @aagonzales! @snidersd @dakahn is that sufficient for WCAG 2.1 compliance? if so then I believe no further action is necessary for our current modal implementations

**Comment by dakahn** [comment]
Sounds like we're good here. 👍
